### PR TITLE
update to proj.4 6.0.0

### PR DESCRIPTION
**Pull request by snowman2**
Connected to issue #152 

This depends on: https://github.com/OSGeo/proj.4/pull/1208
Also, one of the tests fails, but should be fixed with: https://github.com/OSGeo/proj.4/pull/1205

There are a ton of changes - mostly to work with the new API. Feel free to make comments, ask questions, and be thorough in the review.

### CODE REVIEW

**Comment by jswhit** [comment]
Does this really add dependencies on sqlite and cmake?

**Comment by snowman2** [comment]
> Does this really add dependencies on sqlite and cmake?

They have moved from csv files to sqlite database:
"The use of a SQLite-based database for EPSG and other definitions will allow the projects to add more capability (area-aware validation), transition the custom peculiar data structures of the projects to something more universally consumable, and promote definition interoperability between many coordinate system handling software tools."

The cmake dependency makes building proj.4 from source very easy. However, there are other methods for compiling proj.4 without cmake. As aside note, if proj.4 is already installed, cmake is not required.

**Comment by jswhit** [comment]
Maybe this would be a good time to remove the bundled proj4 source, and just require the library as a dependency.  Do your additions to the python API require proj4 6.0.0?

**Comment by snowman2** [comment]
I think the changes I made make it required to have the proj.4 source installed to work. It has a step to compile the source for you in the setup.py internally, but that only works when installing locally from a clone of the repo. Otherwise it results in strange behavior with the path to files that should probably be handled better (could look into relative paths?). Otherwise it requires PROJ_DIR to be set beforehand.

The process definitely could be improved, simplified, or the behavior changed. cartopy is potentially a model that could be followed.

To answer your question, yes it will require version 6.0.0 of proj.4 to work due to the API changes.

**Comment by snowman2** [comment]
@jswhit it looks like the integration with Travis CI has been updated and needs to be changed for the builds to work. To do this, I believe all you need to do is login to Travis CI and activate the `GitHub Apps Integration`.

**Comment by jswhit** [comment]
Hmm. I see the travis tests are passing - so maybe whatever I did to travis actually worked.

File: README.md [hunk 1]
```diff
@@ -3,20 +3,43 @@ pyproj
...
 * clone github repo or download source release at http://python.org/pypi/pyproj.
-  * If you clone the github repo, [Cython](http://cython.org/) is a dependency.
-* python setup.py build
-* python setup.py install (with sudo if necessary).
+  * [Cython](http://cython.org/) is required for the installation.
+
+Setup PROJ.4
+~~~~~~~~~~~~
+
+To install PROJ.4 insternally in pyproj, these are the dependencies:
```
Comment (jswhit) [inline README.md]: type here ('insternally')

File: README.md [hunk 2]
```diff
@@ -3,20 +3,43 @@ pyproj
...
+    * C99
+    * C++11
+    * SQLite 3.7
+    * CMake 3.5
 
-To use proj4 lib (and data files) that are already installed on the system, 
+* git submodule init && git submodule update
+* python install_internal_proj4.py
+
+To use PROJ.4 lib (and data files) that are already installed on the system,
```
Comment (jswhit) [inline README.md]: Does this mean if you have an old proj4 installation, you can still use it (presumably sacrificing the new API you added that uses features from proj4 v6.0.0?)

**Comment by jswhit** [comment]
Build fails with internal proj4 unless I set DYLD_LIBRARY_PATH=$PWD/pyproj/proj_dir/lib.  Even with that I'm getting a failure on import:

ImportError: dlopen(/Users/jsw/Library/Python/2.7/lib/python/site-packages/pyproj-2.0.0-py2.7-macosx-10.11-x86_64.egg/pyproj/_proj.so, 2): Symbol not found: _proj_angular_input
  Referenced from: /Users/jsw/Library/Python/2.7/lib/python/site-packages/pyproj-2.0.0-py2.7-macosx-10.11-x86_64.egg/pyproj/_proj.so

**Comment by snowman2** [comment]
Hmmm, I haven't tested it out on a mac. Maybe adding something on the Travis builds for `osx` would be a good idea.

>This seems to be because the build linked an older version of proj4 on my system, not version 6.0.0.

I added functionality [here](https://github.com/snowman2/pyproj/blob/6463a6fa6066cade6e4a2cf3aba6bb744df83673/setup.py#L25) to let the user know if their version is not compatible. One way to check if that functionality is working would be to change `PROJ_MIN_VERSION` in the setup.py and see if it throws an error based on the version.

Also in the setup.py, the order for searching for PROJ.4 is:
1. PROJ_DIR env variable
2. Check if the internal proj dir exists (pyproj/proj_dir)
3. Check for the `proj` executable on the PATH.

So, you could try tinkering with those options.

File: README.md (hunk 2 above)
Comment (snowman2) [reply README.md]: Unfortunately, no. The proj_api.h is deprecated and the behavior changed within the file. I moved all imports to use the new proj.h. But, wheels and  conda are good solutions for this problem.

Comment (snowman2) [reply README.md]: Well, I take the hard "no" back as I haven't tried it and there might be a way.

Comment (jswhit) [reply README.md]: OK, understood.

**Comment by jswhit** [comment]
I was able to build and install proj.4 separately on macos x, then set PROJ_DIR to point to the installation, installed your gdalbarn pyproj branch and the tests passed.  This is with the specific revision of proj.4 used in the included source - if I use proj.4 master the tests fail.

I'm beginning to think we should just drop the included proj.4 source - there are too many issues building correctly on all platforms.

When is the anticipated release of proj.4 6.0.0?

**Comment by snowman2** [comment]
>if I use proj.4 master the tests fail.

They updated the functionality of their code. I just made some changes so that all should work properly now.

**Comment by jswhit** [comment]
travis is  building proj.4 correctly now, but a couple of tests fail.

**Comment by snowman2** [comment]
I took a look at the tests and it looks like they pushed another change. One of the macos tests failing is unique - `to_epsg`. Is this what you see on your machine?

**Comment by jswhit** [comment]
Yes, that's what I'm seeing on my machine also.

**Comment by snowman2** [comment]
Good to know. I wonder if it has to do with how we are compiling the code or if it is an upstream issue.

**Comment by snowman2** [comment]
I found the bug in the to_epsg function and all is good now.

**Comment by snowman2** [comment]
Looks like I made a blunder when squashing and accidentally squashed one of your commits - sorry about that.

**Comment by snowman2** [comment]
Sounds like a good plan to me. Two thumbs up for using Sphinx 👍

**Comment by snowman2** [comment]
@jswhit if you get the repo connected to Appveyor, I can tinker with getting the tests/builds running there.

**Comment by jswhit** [comment]
Appveyor is enabled now, it should run the next time you make a commit.

**Comment by snowman2** [comment]
The Windows build works! The tests fail, but at least the build works.

**Comment by snowman2** [comment]
@jswhit, I am noticing in the issues that the `preserve_units=False` flag default is throwing some people off. Any issue with me switching the default behavior to be `preserve_units=True` to minimize confusion since this will be a major version change?

**Comment by snowman2** [comment]
Most recent wheel for Python 3.6 (worked alongside an older version of PROJ.4 installed for me) generated on Ubuntu 18.04.1: [pyproj-2.0.0-cp36-cp36m-linux_x86_64.whl.zip](https://github.com/jswhit/pyproj/files/2884900/pyproj-2.0.0-cp36-cp36m-linux_x86_64.whl.zip)

**Comment by jswhit** [comment]
What needs to be done to get the tests passing?

**Comment by snowman2** [comment]
> What needs to be done to get the tests passing?

I think it has to do with the version of pip: https://pip.pypa.io/en/stable/reference/pip/#pep-517-and-518-support. I am pinning to `pip>=10.0.1` and testing.

**Comment by snowman2** [comment]
> worked alongside an older version of PROJ.4 installed for me

For clarification, I have an older version of PROJ.4 on the system and the new version of PROJ.4 (6.0.0) in the wheel does not conflict with it.

**Comment by snowman2** [comment]
The Python 2.7 & 3.6 manylinux wheels build okay, but the 3.4, 3.5, and 3.7 versions are not including the `proj_data` directory for some reason.

**Comment by snowman2** [comment]
All manylinux wheels build properly now.

**Comment by jswhit** [comment]
This is fantastic!

**Comment by jswhit** [comment]
Looks like we will be able to merge this and make a new release when proj4 6.0.0 is released (supposedly March 1?)

**Comment by jswhit** [comment]
The README should probably mention that proj4 version 6.0.0 is required.

**Comment by snowman2** [comment]
@jswhit I definitely hope that pyproj can be released along with the next version of PROJ.4.

What are your thoughts about:
- [ ] Using [coveralls](https://coveralls.io/)
- [ ] Using pytest for testing instead of nose2
- [ ] Switching to Sphinx for docs (you mentioned you may be interested above)

An example .travis.yml using pytest with coveralls is [here](https://github.com/snowman2/gazar/blob/7e0317466afa96ccdd5dd214b924b9c93d8a7788/.travis.yml#L56).

**Comment by jswhit** [comment]
I don't have an opinion on coveralls and pytest - I'll defer to those who know more about testing frameworks.  Switching to sphinx for docs really needs to be done before the release.

**Comment by snowman2** [comment]
Sounds like a good plan to me. The test framework currently in use works fine, so a decision on that can be postponed to after the release. Would it be helpful for me to to add Sphinx docs to this PR?

**Comment by jswhit** [comment]
Great! I think this is ready to merge now - what do you think @snowman2 and @micahcochran?   After merging, I'll create a ticket for sphinx doc updates which should be done before a release.

**Comment by snowman2** [comment]
@jswhit, sounds good to me. No issues here. Do you have access to the windows wheels uploaded as artifacts in Appveyor?

**Comment by snowman2** [comment]
Found them:

https://ci.appveyor.com/api/buildjobs/235jw58mg0qsi9xr/artifacts/dist%2Fpyproj-2.0.0-cp35-cp35m-win_amd64.whl
https://ci.appveyor.com/api/buildjobs/964dt9t1n9d27709/artifacts/dist%2Fpyproj-2.0.0-cp36-cp36m-win_amd64.whl
https://ci.appveyor.com/api/buildjobs/so47mvw7a53a7lsm/artifacts/dist%2Fpyproj-2.0.0-cp37-cp37m-win_amd64.whl

I am curious how they work. Need to test on a windows machine.
Should this release drop support for Python 2.7 since it it dies this year? Or do you want to add the Python 2.7 build back in appveyor?

**Comment by snowman2** [comment]
The windows wheels are faulty. I have some ideas to try if you don't mind holding off merging for a bit.

**Comment by snowman2** [comment]
The wheels with the proj.lib & sqlite3.dll aren't working either. Not sure why. Well, feel free to merge anyway and maybe an issue can be made to look into it later.

**Pull request merged**